group/assignment6/ParkingLotChallenge3.java

```diff
@@ -0,0 +1,260 @@
+                            nextLoopWaitingSpaces.add(prevSpot); //it no longer contains its end car
+                        } else {
+                            continue;
+                        }
+                        break;
+                    }
+                }
+                if (movingCar == -1) {
+                    throw new Exception("Out of options, constraints make the moves unpossible.");
+                }
+
+                //test the new state
+                waitingSpaces = nextLoopWaitingSpaces;
+            } else {
+                done = true;
+            }
+        }
+
+    }
+
+    private void doAllMoves() {
+
+        // Iterate through and check whether current and end state match.
+        // While possible, reduce the number of moves by trying to move to the current free space the car that should be there in the end.
+        // When the current free spot matches the end free spot, swap the the car at position i with the car that is supposed to be at that
+        // position by making two moves. This will change the position of the free space again and trigger the while loop in the next iteration
+        // minimizing the number of moves need to rearrage the cars at the parking lot.
+        for (int i = 0; i < parkingLotSize; i++) {
+            // If possible, firstly move to the current free space the car that should be there in the end.
+            // this while reduces the number of moves from 12 to 6 for testPermutation
+            // and from 5 to 3 in testSimpleCase.
+            while (currentEmptySpotIndex != endEmptySpotIndex) {
+                int desiredCar = parkingLotEnd[currentEmptySpotIndex];
+                int currentDesiredCarPosition = carToPosition[desiredCar];
+                doOneMove(desiredCar, currentDesiredCarPosition, currentEmptySpotIndex);
+            }
+
+            int currentCar = parkingLotCurrent[i];
+            int desiredCar = parkingLotEnd[i];
+            if (currentCar != desiredCar) {
+                // move from the space, only if it's not already empty
+                if (currentCar != -1) {
+                    doOneMove(currentCar, i, currentEmptySpotIndex);
+                }
+                // move to the space, only if it should not be empty
+                if (desiredCar != -1) {
+                    int desiredCarPosition = carToPosition[desiredCar];
+                    doOneMove(desiredCar, desiredCarPosition, i);
+                }
+            }
+        }
+    }
+
+    private void doOneMove(int carId, int fromSpace, int toSpace) {
+        moves.add(new Move(carId, fromSpace, toSpace));
+        parkingLotCurrent[toSpace] = carId;
+        parkingLotCurrent[fromSpace] = -1;
+        currentEmptySpotIndex = fromSpace;
+        carToPosition[carId] = toSpace;
+
+        //add the new state to the previous states
+        previousStates.add(parkingLotCurrent);

```

Comment:
I think previousStates is not initialized before so this statement returns null pointer exception.